lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -253,51 +435,360 @@ void FileEmitter::emitInstance(llvm::json::OStream &json,
+          .Case<comb::ICmpOp>([](auto cmpOp) -> StringRef {
+            switch (cmpOp.getPredicate()) {
+            case comb::ICmpPredicate::eq:
+              return "==";
+            case comb::ICmpPredicate::ne:
+              return "!=";
+            case comb::ICmpPredicate::ceq:
+              return "===";
+            case comb::ICmpPredicate::cne:
+              return "!==";
+            case comb::ICmpPredicate::weq:
+              return "==?";
+            case comb::ICmpPredicate::wne:
+              return "!=?";
+            case comb::ICmpPredicate::ult:
+            case comb::ICmpPredicate::slt:
+              return "<";
+            case comb::ICmpPredicate::ugt:
+            case comb::ICmpPredicate::sgt:
+              return ">";
+            case comb::ICmpPredicate::ule:
+            case comb::ICmpPredicate::sle:
+              return "<=";
+            case comb::ICmpPredicate::uge:
+            case comb::ICmpPredicate::sge:
+              return ">=";
+            }
+            return {};
+          })
+          .Default([](auto) { return ""; });
+  if (!binaryOpcode.empty() && op->getNumOperands() == 2) {
+    auto lhs = emitExpression(op->getOperand(0));
+    auto rhs = emitExpression(op->getOperand(1));
+    if (!lhs || !rhs)
+      return {};
+    return {JObject({
+                {"opcode", binaryOpcode},
+                {"operands", {lhs.expr, rhs.expr}},
+            }),
+            result.getType()};
+  }
+
+  // Expand variadic combinational ops into nested binary HGLDD operations.
+  StringRef variadicOpcode = TypeSwitch<Operation *, StringRef>(op)
+                                 .Case<comb::AndOp>([](auto) { return "&"; })
+                                 .Case<comb::OrOp>([](auto) { return "|"; })
+                                 .Case<comb::XorOp>([](auto) { return "^"; })
+                                 .Default([](auto) { return ""; });
+  if (!variadicOpcode.empty()) {
+    auto operands = op->getOperands();
+    auto value = emitExpression(operands[0]);
+    if (!value)
+      return {};
+    operands = operands.drop_front();
+    while (!operands.empty()) {
+      auto otherValue = emitExpression(operands[0]);
+      if (!otherValue)
+        return {};
+      operands = operands.drop_front();
+      value = {JObject({
+                   {"opcode", variadicOpcode},
+                   {"operands", {value.expr, otherValue.expr}},
+               }),
+               result.getType()};
     }
+    return value;
   }
 
-  json.objectEnd();
+  // Special handling for concatenation.
+  if (auto concatOp = dyn_cast<comb::ConcatOp>(op)) {
+    std::vector<JValue> args;
+    for (auto operand : concatOp.getOperands()) {
+      auto value = emitExpression(operand);
+      if (!value)
+        return {};
+      args.push_back(value.expr);
+    }
+    return {JObject({
+                {"opcode", "{}"},
+                {"operands", args},
```

Comment:
This `JObject({"opcode", opcode}, {"operands", args}})` seems to be pretty common, can we create a helper function to avoid writing raw strings "opcode" and  "operands"?

Reply:
Yeah good idea!